# Сингулярное разложение матриц

In [1]:
import numpy as np

1. Найти с помощью NumPy SVD для матрицы

$$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}.$$

Представим матрицу $A = \begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}$ в виде

### $$A=UDV^{T}$$

In [2]:
A = np.array([[1, 2, 0], [0, 0, 5], [3, -4, 2], [1, 6, 5], [0, 1, 0]])

U, s, W = np.linalg.svd(A)

# Транспонируем матрицу W
V = W.T

# s - список диагональных элементов, его нужно привести к виду диагональной матрицы для наглядности
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [3]:
print(f'Матрица U:\n{U}')

Матрица U:
[[ 0.17056501  0.15680918 -0.53077508 -0.79905375 -0.16158397]
 [ 0.39287016 -0.52933945  0.6134793  -0.43375771  0.03082495]
 [-0.14366152 -0.82449256 -0.52379105  0.14049848  0.07400343]
 [ 0.88843702  0.06074346 -0.24655277  0.37755832 -0.06042632]
 [ 0.08125046  0.10831843 -0.08231425 -0.10524851  0.98173958]]


In [4]:
print(f'Матрица D:\n{D}')

Матрица D:
[[8.82486885 0.         0.        ]
 [0.         6.14060608 0.        ]
 [0.         0.         2.53271528]
 [0.         0.         0.        ]
 [0.         0.         0.        ]]


In [5]:
print(f'Матрица V:\n{V}')

Матрица V:
[[ 0.07116451 -0.36737824 -0.92734505]
 [ 0.71702467  0.66514082 -0.20847855]
 [ 0.69340553 -0.65009301  0.31075368]]


2. Для матрицы из предыдущего задания найти:

    а) евклидову норму;
    
    б) норму Фробениуса.

a) Евклидова норма матрицы равна евклидовой норме диагональной матрицы из ее сингулярных чисел $D$:
### $$\left \| A \right \|_{E}^{2}=\text{max}\left (\frac{(Ax,Ax)}{(x,x)}\right )=\text{max}\left (\frac{(UDV^{T}x,UDV^{T}x)}{(V^{T}x,V^{T}x)}\right )=\text{max}\left (\frac{(Dz,Dz)}{(z,z)}\right ).$$
Максимальное значение этого отношения будет равно максимальному сингулярному числу $\mu_{max}$. Принимая во внимание факт невозрастания сингулярных чисел, находящихся на диагонали матрицы $D$, получаем, что
### $$\left \| A \right \|_{E}=\mu_{1} \approx 8.8249$$

In [6]:
np.linalg.norm(A, 2)

8.824868854820444

б) Используя найденное ранее сингулярное разложение матрицы $A$, норму Фробениуса можно вычислить следующим образом:
### $$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{3}\mu_{k}^{2}}$$
### $$\left \| A \right \|_{F}=\sqrt{8.8249^2 + 6.1406^2 + 2.5327^2} \approx 11.0454$$

In [7]:
np.sqrt(sum(s**2))

11.045361017187263

In [8]:
np.linalg.norm(A)

11.045361017187261